Allow the exp claim in IntrospectionResponse to be passed as a string #3202

ovindu-a opened the pull request:
## Purpose

This allows the exp value in IntrospectionResponse to have string value as mentioned in [issue #3202](https://github.com/ballerina-platform/ballerina-library/issues/3202). I have changed the code in the `prepareIntrospectionResponse` function to check if the exp value passed in as json is a string and if so it tries to cast it to and int and assign it. 

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ballerina-platform/module-ballerina-oauth2?pullRequest=1314) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ballerina-platform/module-ballerina-oauth2?pullRequest=1314) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ballerina-platform/module-ballerina-oauth2?pullRequest=1314) it.</sub>

MohamedSabthar commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -210,7 +210,22 @@ isolated function prepareIntrospectionResponse(json payload) returns Introspecti
             }
             EXP => {
-                introspectionResponse.exp = <int>payloadMap[key];
+                anydata value = payloadMap[key];
+
+                if value is int {
+                    introspectionResponse.exp = value;
+                } else if value is string {
+                    int|error parsedInt = int:fromString(value);
+                    if parsedInt is int {
+                        introspectionResponse.exp = parsedInt;
+                    } else {
+                        introspectionResponse.exp = ();
+                        log:printError("Failed to parse string to integer for exp field", parsedInt);
+                    }
+                } else {
+                    introspectionResponse.exp = ();
+                    log:printError("Invalid type for exp field, expected int or string");
+                }
```
```suggestion
                json value = payloadMap[key];
                if value is int {
                    introspectionResponse.exp = value;
                } else if value is string {
                    int|error parsedInt = int:fromString(value);
                    if parsedInt is int {
                        introspectionResponse.exp = parsedInt;
                    } else {
                        introspectionResponse.exp = ();
                        log:printError("Failed to parse string to integer for exp field", parsedInt);
                    }
                } else {
                    introspectionResponse.exp = ();
                    log:printError("Invalid type for exp field, expected int or string");
                }
```

Logic looks good. Lets move this logic into a separate function. 
```
isolated function parseExpClaim(json expClaim) returns int? {
     // logic
}
``

MohamedSabthar replied:
Add a test for the `prepareIntrospectionResponse` covering the logic you have implemented.

MohamedSabthar commented:
@ovindu-a , your solution looks good. I've added a few comments. Let's address them. Please remove any unnecessary formatting changes and keep the PR focused on the implemented logic. Let's also change the PR title to something more descriptive.

ovindu-a commented:
Noted @MohamedSabthar, I will make the changes

ovindu-a commented:
@MohamedSabthar I have made 2 commits
1. Extracted the parsing code to a function. Also removed the formatting that seems to have been added automatically.
2. Added a test case to check if a string value is parsed to int.

Please let me know if there are anymore changes

MohamedSabthar commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -210,7 +210,7 @@ isolated function prepareIntrospectionResponse(json payload) returns Introspecti
             }
             EXP => {
-                introspectionResponse.exp = <int>payloadMap[key];
+                introspectionResponse.exp = parseExpClaim(<json>payloadMap[key]);
```
```suggestion
                introspectionResponse.exp = parseExpClaim(payloadMap[key]);
```

MohamedSabthar commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -278,3 +278,20 @@ isolated function validateFromCache(cache:Cache oauth2Cache, string token) retur
+        log:printError("Invalid type for exp field, expected int or string");
+        return ();
+    }
+}
```
```suggestion
}

```

ovindu-a replied:
Was this a mistake or is this change necessary? @MohamedSabthar

MohamedSabthar replied:
Need to end each file with a new line @ovindu-a

ovindu-a replied:
Ohh right

MohamedSabthar commented on ballerina/tests/listener_oauth2_provider_test.bal:
```diff
@@ -545,3 +545,24 @@ isolated function testTokenIntrospectionRequestWithSecureSocketAndWithHttpUrlSch
 }
+
+// Test that the `exp` field is correctly parsed as an integer when given an example string value for `exp` field
+@test:Config {}
+isolated function testStringExpParsing() {
+    json simulatedResponse = {
+        "active": true,
+        "exp": "1672531199",
+        "scope": "read write",
+        "client_id": "test_client_id",
+        "username": "test_user"
+    };
+
+    IntrospectionResponse response = prepareIntrospectionResponse(simulatedResponse);
+
+
+    test:assertTrue(response.active);
+    test:assertEquals(response.exp, 1672531199);
+    test:assertEquals(response.scope, "read write");
+    test:assertEquals(response.clientId, "test_client_id");
+    test:assertEquals(response.username, "test_user");
+}
```
```suggestion
@test:Config {}
isolated function testPrepareIntrospectionResponseWithStringExpClaim() {
    json simulatedResponse = {
        active: true,
        exp: "1672531199",
        scope: "read write",
        client_id: "test_client_id",
        username: "test_user"
    };
    IntrospectionResponse response = prepareIntrospectionResponse(simulatedResponse);
    test:assertEquals(response.exp, 1672531199);
}
```
No need to test other fields, just validating the exp claim is enough here, and let's change the function name to suggested one or more descriptive one.

MohamedSabthar commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -278,3 +278,20 @@ isolated function validateFromCache(cache:Cache oauth2Cache, string token) retur
+isolated function parseExpClaim(json expClaim) returns int|() {
+    if expClaim is int {
+        return expClaim;
+    } else if expClaim is string {
```
```suggestion
    } 
    if expClaim is string {
```

MohamedSabthar commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -278,3 +278,20 @@ isolated function validateFromCache(cache:Cache oauth2Cache, string token) retur
+        if parsedInt is int {
+            return parsedInt;
+        } else {
+            log:printError("Failed to parse string to integer for exp field", parsedInt);
```
```suggestion
            log:printError("Failed to parse the `exp` claim value to a valid integer.", parsedInt);
```

MohamedSabthar commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -278,3 +278,20 @@ isolated function validateFromCache(cache:Cache oauth2Cache, string token) retur
     return;
 }
+
+isolated function parseExpClaim(json expClaim) returns int|() {
+    if expClaim is int {
+        return expClaim;
+    } else if expClaim is string {
+        int|error parsedInt = int:fromString(expClaim);
+        if parsedInt is int {
+            return parsedInt;
+        } else {
+            log:printError("Failed to parse string to integer for exp field", parsedInt);
+            return ();
+        }
+    } else {
+        log:printError("Invalid type for exp field, expected int or string");
+        return ();
+    }
```
```suggestion
isolated function parseExpClaim(json expClaim) returns int|() {
    if expClaim is string {
        return checkpanic int:fromString(expClaim);
    } 
    return <int>expClaim;
```
Instead of logging the error let's do this @ovindu-a.

MohamedSabthar commented:
> @MohamedSabthar I have made 2 commits
> 
> 1. Extracted the parsing code to a function. Also removed the formatting that seems to have been added automatically.
> 2. Added a test case to check if a string value is parsed to int.
> 
> Please let me know if there are anymore changes

Added a few more suggestions, @ovindu-a. Let's address them and get this merged.

ovindu-a commented:
I have resolved the issues mentioned in the most recent commit. @MohamedSabthar

ThisaruGuruge commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -260,7 +260,7 @@ isolated function validateFromCache(cache:Cache oauth2Cache, string token) retur
     }
     if cachedEntry is any {
-        IntrospectionResponse response = <IntrospectionResponse> cachedEntry;
+        IntrospectionResponse response = <IntrospectionResponse>cachedEntry;
```
Can't we use this:

```ballerina
isolated function cacheType(cache:Cache oauth2Cache, string token) returns oauth2:IntrospectionResponse? {
    oauth2:IntrospectionResponse|error cachedEntry = oauth2Cache.get(token).ensureType();
    if cachedEntry is oauth2:IntrospectionResponse {
        return cachedEntry;
    }
    return;
}
```

ovindu-a replied:
This function will not check the exp claim to see if it is expired right?

MohamedSabthar replied:
This seems to be a formatting change. However I have created this issue: https://github.com/ballerina-platform/ballerina-library/issues/7237 to refactor the codebase with best practices. We can handle this separately.

ThisaruGuruge commented on ballerina/listener_oauth2_provider.bal:
```diff
@@ -278,3 +278,10 @@ isolated function validateFromCache(cache:Cache oauth2Cache, string token) retur
+
+isolated function parseExpClaim(json expClaim) returns int|() {
+    if expClaim is string {
+        return checkpanic int:fromString(expClaim);
```
There are a few concerns here.

1. Using `checkpanic` is not recommended. 
2. We should handle the errors correctly.

MohamedSabthar replied:
Yes, using `checkpanic` is not recommended, but the current implementation panics in several places due to casting, primarily in the `prepareIntrospectionResponse` method. Ideally, these cases should use the ensureType method, and on error, the introspection response should be considered invalid. As a result, OAuth2 validation should fail since the authorization server does not comply with the RFC specification. I have created a separate issue for this: https://github.com/ballerina-platform/ballerina-library/issues/7231, and we can address it with a new PR, including this checkpanic case.

ovindu-a replied:
So any changes to be made on this? @MohamedSabthar

ThisaruGuruge replied:
@MohamedSabthar +1

MohamedSabthar replied:
> So any changes to be made on this? @MohamedSabthar
 
Will handle it in a separate PR.
